math/src/elliptic_curve/short_weierstrass/point.rs

```diff
@@ -377,16 +383,226 @@ where
+            let z3 = (y1 * z1).double(); // Z3 = 2 * Y1 * Z1
+
+            Self::new([x3, y3, z3])
+        } else {
+            // http://www.hyperelliptic.org/EFD/g1p/data/shortw/jacobian-0/doubling/dbl-2009-alnr
+            // http://www.hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-0.html#doubling-dbl-2009-l
+            let xx = x1.square(); // XX = X1^2
+            let yy = y1.square(); // YY = Y1^2
+            let yyyy = yy.square(); // YYYY = YY^2
+            let zz = z1.square(); // ZZ = Z1^2
+            let s = ((x1 + &yy).square() - &xx - &yyyy).double(); // S = 2 * ((X1 + YY)^2 - XX - YYYY)
+            let m = &xx.double() + &xx + &a_coeff * &zz.square(); // M = 3 * XX + a * ZZ^2
+            let x3 = m.square() - &s.double(); // X3 = M^2 - 2 * S
+            let y3 = m * (&s - &x3) - &yyyy.double().double().double(); // Y3 = M * (S - X3) - 8 * YYYY
+            let z3 = (y1 + z1).square() - &yy - &zz; // Z3 = (Y1 + Z1)^2 - YY - ZZ
+
+            Self::new([x3, y3, z3])
+        }
+    }
+
+    pub fn operate_with_affine(&self, other: &Self) -> Self {
+        let [x1, y1, z1] = self.coordinates();
+        let [x2, y2, z2] = other.coordinates();
+
+        if *z2 != FieldElement::one() {
+            panic!("The second point must be in affine coordinates (Z = 1)");
+        }
+
+        if self.is_neutral_element() {
+            return other.clone();
+        }
+        if other.is_neutral_element() {
+            return self.clone();
+        }
+
+        let z1z1 = z1.square();
+        let u1 = x1;
+        let u2 = x2 * &z1z1;
+        let s1 = y1;
+        let s2 = y2 * z1 * &z1z1;
+        let h = &u2 - u1;
+        let hh = h.square();
+        let hhh = &h * &hh;
+        let r = &s2 - s1;
+        let v = u1 * &hh;
+        let x3 = r.square() - &hhh - &v - &v;
+        let y3 = r * (&v - &x3) - s1 * &hhh;
+        let z3 = z1 * &h;
+
+        Self::new([x3, y3, z3])
+    }
+}
+
+impl<E: IsEllipticCurve> PartialEq for ShortWeierstrassJacobianPoint<E> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl<E: IsEllipticCurve> Eq for ShortWeierstrassJacobianPoint<E> {}
+
+impl<E: IsShortWeierstrass> FromAffine<E::BaseField> for ShortWeierstrassJacobianPoint<E> {
+    fn from_affine(
+        x: FieldElement<E::BaseField>,
+        y: FieldElement<E::BaseField>,
+    ) -> Result<Self, crate::elliptic_curve::traits::EllipticCurveError> {
+        if E::defining_equation(&x, &y) != FieldElement::zero() {
+            Err(EllipticCurveError::InvalidPoint)
+        } else {
+            let coordinates = [x, y, FieldElement::one()];
+            Ok(ShortWeierstrassJacobianPoint::new(coordinates))
+        }
+    }
+}
+
+impl<E: IsShortWeierstrass> IsGroup for ShortWeierstrassJacobianPoint<E> {
+    /// The point at infinity.
+    fn neutral_element() -> Self {
+        Self::new([
+            FieldElement::zero(),
+            FieldElement::one(),
```

Comment:
The neutral element for jacobian coordinates is different. It is either (1,1,0) or (1,-1,0)